cpu/sam0_common/periph/usbdev.c

```diff
@@ -242,8 +242,41 @@ static inline void _poweron(void)
                                    GCLK_CLKCTRL_GEN_GCLK0 |
                                    (GCLK_CLKCTRL_ID(USB_GCLK_ID)));
 #elif defined(CPU_FAM_SAML21)
+    /* enable DFLL */
+    OSCCTRL->DFLLCTRL.reg = OSCCTRL_DFLLCTRL_ENABLE;
+    while (!(OSCCTRL->STATUS.reg & OSCCTRL_STATUS_DFLLRDY)) {}
+
+    /* setup initial values for closed loop */
+    OSCCTRL->DFLLVAL.reg = OSCCTRL_DFLLVAL_COARSE((*(uint32_t*)NVMCTRL_OTP5)
+                           >> 26) |  OSCCTRL_DFLLVAL_FINE(512);
+    while (!(OSCCTRL->STATUS.reg & OSCCTRL_STATUS_DFLLRDY)) {}
+
+    /* we will generate a 48 Mhz clock from 1KHz SOF signal */
+    OSCCTRL->DFLLMUL.reg = OSCCTRL_DFLLMUL_CSTEP(0x08) |
+                           OSCCTRL_DFLLMUL_FSTEP(0x08) |
+                           OSCCTRL_DFLLMUL_MUL((48000000U/1000));
+
+    /* Disable DFLL before setting its configuration */
+    OSCCTRL->DFLLCTRL.reg = 0;
+    while (!(OSCCTRL->STATUS.reg & OSCCTRL_STATUS_DFLLRDY)) {}
+
+    /* Configure DFLL for USB clock recovery mode */
+    OSCCTRL->DFLLCTRL.reg =  OSCCTRL_DFLLCTRL_USBCRM |
+                             OSCCTRL_DFLLCTRL_MODE |
+                             OSCCTRL_DFLLCTRL_CCDIS |
+                             OSCCTRL_DFLLCTRL_ENABLE;
+
+    /* Ensure COARSE and FINE are locked */
+    while((!(OSCCTRL->STATUS.bit.DFLLLCKC)) && (!(OSCCTRL->STATUS.bit.DFLLLCKF))) {}
+    while (!(OSCCTRL->STATUS.bit.DFLLRDY)) {}
+
+    /* set clock 1 to 48 MHz DFLL */
+    GCLK->GENCTRL[1].reg = GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_DFLL48M;
+    while (GCLK->SYNCBUSY.reg & GCLK_SYNCBUSY_GENCTRL(1)) {}
```

Comment:
I'd much rather see that DFLL setup code in `saml21/cpu.c`.
It isn't related to usbdev - that just happens to use that clock.

If we want to use the DFLL for something else in the future, we better not duplicate the setup code.

samd5x does that too - set up the clock in `cpu_init()` if `MODULE_PERIPH_USBDEV` is used, then the user (usbdev) just connects to the existing GCLK. This also has the benefit that all GCLKs are set up in one place.

Reply:
My use case for keeping the clock initialization separate is a energy harvesting node. I would like to carry out configuration via USB, but the 48MHz clock should stay off unless needed, in order to save power.

Reply:
You can connect all the peripheral to a separate GCLK that stays constant, then you can change the main clock independently.

But yes, the whole GCLK system needs some proper overhaul in RIOT.

Something like `gclk_connect()` and `gclk_disconnect()` and then disabling the clock of there are no more users should definitely be part of it.